R/scale_marker.R

```diff
@@ -0,0 +1,44 @@
+
+
+scale_marker_discrete <- function(glyphs = NULL, colors = NULL, limits = NULL, ...) {
+
+  markers <- data.frame(glyphs = glyphs, colors = colors, labels = limits) |>
+    distinct()
+
+  palette <- pal_markers(
+    glyphs = markers$glyphs,
+    colors = markers$colors,
+    n_values = nrow(markers)
+  )
+
+  discrete_scale("marker", rlang::missing_arg(), palette = palette,
+                 limits = markers$labels, ..., na.translate = FALSE)
+}
+
+
+.default_glyphs <- c("⛔", "✔", "✖", "☀", "☢", "☠", "☮", "⚛", "♥",
+                     "✀")
+
+pal_markers <- function(glyphs = NULL, colors = NULL, n_values = NULL) {
+
+  n_values <- n_values %||% max(length(glyphs), length(colors))
+  if (n_values == 0) n_values <- length(.default_glyphs)
+
+  markers <- vctrs::new_rcrd(
+    list(
+      colour = rep(colors %||% "black", length.out = n_values),
+      glyphs = rep(glyphs %||% .default_glyphs, length.out = n_values)
+    ),
+    class = "marker"
+  )
```

Comment:
We're essentially making a 2-column data frame act like a vector, so that it works well with ggplot2's internals.
We unpack this in the drawing method.

Reply:
@teunbrand I've been working through this example to wrap my head around {vctrs} and how `scale_marker_discrete()` is triggered during the call to `geom_swim_marker()`.

I think I have a shallow handle of {vctrs} here, but wanted to test it by converting the code above to something I could more easily inspect:

```r
  markers <- structure(
    list(
      colour = rep(colors %||% "black", length.out = n_values),
      glyphs = rep(glyphs %||% .default_glyphs, length.out = n_values)
    ),
    class = "marker"
  )
  ```
I believe this should work the same, but on stepping through a `geom_swim_marker()` call it seems this set up triggers `draw_panel` before it triggers `draw_key` and winds up failing. Any insight as to what I'm missing here in how this behaves?

Is it also correct to assume that the definition of "marker" in `required_aes` triggers ggplot to look for `scale_*_discrete()` (in this case `scale_marker_discrete()`)?

Reply:
Yeah so `vctrs::new_rcrd()` invokes the <vctrs_rcrd> class that has all sorts of niceties that makes the data-frame like structure behave as if it was a vector (by treating each row as element). It avoids some type unsafeties in the internals of ggplot2 that would mangle the structure otherwise.

Reply:
Very cool to learn. Thanks!